### PR TITLE
[pornflip] Add PornFlip extractor

### LINKED ISSUE #11556: Error downloading https://www.pornflip.com/v/nk7o8atTY4Q
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like that [x])
- Use *Preview* tab to see how your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2016.12.22*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [X] I've **verified** and **I assure** that I'm running youtube-dl **2016.12.22**

### Before submitting an *issue* make sure you have:
- [X] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [X] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---
$ youtube-dl https://www.pornflip.com/v/nk7o8atTY4Q --verbose
[debug] System config: []
[debug] User config: ['-o', 'G:\\Images\\Vids\\%(title)s.%(ext)s', '--restrict-filenames', '--format', 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best', '--no-p
laylist', '--no-mtime', '--username', 'PRIVATE', '--password', 'PRIVATE', '--write-info-json', '--no-overwrites', '--autonumber-size', '3']
[debug] Command-line args: ['https://www.pornflip.com/v/nk7o8atTY4Q', '--verbose']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2016.12.22
[debug] Python version 3.4.4 - Windows-10-10.0.14393
[debug] exe versions: none
[debug] Proxy map: {}
[generic] nk7o8atTY4Q: Requesting header
WARNING: Falling back on generic information extractor.
[generic] nk7o8atTY4Q: Downloading webpage
[generic] nk7o8atTY4Q: Extracting information
[generic] nk7o8atTY4Q: Downloading MPD manifest
ERROR: Failed to download MPD manifest: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the
 latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmptqqgo4yk\build\youtube_dl\extractor\common.py", line 405, in _request_webpage
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmptqqgo4yk\build\youtube_dl\YoutubeDL.py", line 2001, in urlopen
  File "C:\Python\Python34\lib\urllib\request.py", line 470, in open
  File "C:\Python\Python34\lib\urllib\request.py", line 580, in http_response
  File "C:\Python\Python34\lib\urllib\request.py", line 508, in error
  File "C:\Python\Python34\lib\urllib\request.py", line 442, in _call_chain
  File "C:\Python\Python34\lib\urllib\request.py", line 588, in http_error_default---

### PR DESCRIPTION
**Pull request by einstein95**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds an extractor for pornflip.com. This fixes #11556 albeit with a non-copyrighted video (afaik).

### CODE REVIEW

File: youtube_dl/extractor/pornflip.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                formats.append(f)
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'formats': formats,
+            'title': flashvars['video_vars[title]'][0],
+            'uploader': uploader,
+            'thumbnail': flashvars['video_vars[big_thumb]'][0],
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Sholdn not break the extraction if missing.

File: youtube_dl/extractor/pornflip.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            'age_limit': '18',
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Must be numeric.

File: youtube_dl/extractor/pornflip.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    'url': url
+                }
+                formats.append(f)
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'formats': formats,
+            'title': flashvars['video_vars[title]'][0],
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: This should be extracted first.

File: youtube_dl/extractor/pornflip.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        webpage = self._download_webpage(url, video_id)
+        uploader = self._html_search_regex(
+            r'<span class="name">\s+<a class="ajax" href=".+>\s+<strong>([^<]+)<', webpage, 'uploader', fatal=False)
+        flashvars = compat_parse_qs(self._html_search_regex(
+            r'<embed.+?flashvars="([^"]+)"',
+            webpage, 'flashvars'))
+        formats = []
+        for flashvar in flashvars.keys():
+            if flashvar.startswith('video_vars[video_urls]'):
+                quality = self._search_regex(r'video_vars\[video_urls\]\[(.+)\]', flashvar, 'quality')
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Should not break the extraction if missing.

File: youtube_dl/extractor/pornflip.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                url = flashvars[flashvar][0]
+                f = {
+                    'quality': quality,
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Must be numeric.

**Comment by einstein95** [comment]
@dstftw Fixes have been made

File: youtube_dl/extractor/pornflip.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            webpage, 'flashvars'))
+        title = flashvars['video_vars[title]'][0]
+        try:
+            thumbnail = flashvars['video_vars[big_thumb]'][0]
+        except KeyError:
+            thumbnail = ''
+        formats = []
+        for flashvar in flashvars.keys():
+            if flashvar.startswith('video_vars[video_urls]'):
+                quality = self._search_regex(r'video_vars\[video_urls\]\[(\d+).+?\]', flashvar, 'quality')
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Nothing changed. Breaks in case regex does not match.

File: youtube_dl/extractor/pornflip.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        uploader = self._html_search_regex(
+            r'<span class="name">\s+<a class="ajax" href=".+>\s+<strong>([^<]+)<', webpage, 'uploader', fatal=False)
+        flashvars = compat_parse_qs(self._html_search_regex(
+            r'<embed.+?flashvars="([^"]+)"',
+            webpage, 'flashvars'))
+        title = flashvars['video_vars[title]'][0]
+        try:
+            thumbnail = flashvars['video_vars[big_thumb]'][0]
+        except KeyError:
+            thumbnail = ''
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: `utils.try_get`.

File: youtube_dl/extractor/pornflip.py (hunk 6 above)
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: This field is height not quality.

File: youtube_dl/extractor/pornflip.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            if flashvar.startswith('video_vars[video_urls]'):
+                quality = self._search_regex(r'video_vars\[video_urls\]\[(\d+).+?\]', flashvar, 'quality')
+                if quality = '':
+                    continue
+                url = flashvars[flashvar][0]
+                f = {
+                    'quality': int(quality),
+                    'url': url
+                }
+                formats.append(f)
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Append dict right away.

File: youtube_dl/extractor/pornflip.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        flashvars = compat_parse_qs(self._html_search_regex(
+            r'<embed.+?flashvars="([^"]+)"',
+            webpage, 'flashvars'))
+        title = flashvars['video_vars[title]'][0]
+        try:
+            thumbnail = flashvars['video_vars[big_thumb]'][0]
+        except KeyError:
+            thumbnail = ''
+        formats = []
+        for flashvar in flashvars.keys():
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: `for k, v in flashvars.items()`.

File: youtube_dl/extractor/pornflip.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            if flashvar.startswith('video_vars[video_urls]'):
+                quality = self._search_regex(r'video_vars\[video_urls\]\[(\d+).+?\]', flashvar, 'quality')
+                if quality = '':
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Invalid syntax at all.

**Comment by einstein95** [comment]
@dstftw Done, didn't see that syntax one. (Happens to me more times than I'd care to admit.)

**Comment by einstein95** [comment]
@dstftw Done

File: youtube_dl/extractor/pornflip.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        uploader = self._html_search_regex(
+            r'<span class="name">\s+<a class="ajax" href=".+>\s+<strong>([^<]+)<', webpage, 'uploader', fatal=False)
+        flashvars = compat_parse_qs(self._html_search_regex(
+            r'<embed.+?flashvars="([^"]+)"',
+            webpage, 'flashvars'))
+        title = flashvars['video_vars[title]'][0]
+        thumbnail = try_get(flashvars, lambda x: x['video_vars[big_thumb]'])[0]
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: `[0]` must be in `getter`.

File: youtube_dl/extractor/pornflip.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        formats = []
+        print(flashvars)
+        for k, v in flashvars.items():
+            if k.startswith('video_vars[video_urls]'):
+                try:
+                    height = self._search_regex(r'video_vars\[video_urls\]\[(\d+).+?\]', k, 'height')
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Nothing changed.

**Comment by einstein95** [comment]
@dstftw Done

File: youtube_dl/extractor/pornflip.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+            webpage, 'flashvars'))
+        title = flashvars['video_vars[title]'][0]
+        thumbnail = try_get(flashvars, lambda x: x['video_vars[big_thumb]'][0])
+        formats = []
+        for k, v in flashvars.items():
+            if k.startswith('video_vars[video_urls]'):
+                try:
+                    height = self._search_regex(r'video_vars\[video_urls\]\[(\d+).+?\]', k, 'height')
+                except RegexNotFoundError:
+                    continue
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: Use `default=None` to `_search_regex` instead. `k.startswith('video_vars[video_urls]')` check is redundant.

File: youtube_dl/extractor/pornflip.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+                url = flashvars[k][0]
```
Comment (dstftw) [inline youtube_dl/extractor/pornflip.py]: `flashvars[k]` is `v`.

**Comment by dstftw** [comment]
Squash commits.

**Comment by einstein95** [comment]
@dstftw Done and squashed.

**Pull request closed without merging**